include/openthread/dns_client.h

```diff
@@ -0,0 +1,470 @@
+ * addresses. Index zero gets the first address and so on. When we reach end of the list, `OT_ERROR_NOT_FOUND` is
+ * returned.
+ *
+ * @param[in]  aResponse     A pointer to the response.
+ * @param[in]  aIndex        The address record index to retrieve.
+ * @param[out] aAddress      A pointer to a IPv6 address to output the address (MUST NOT be NULL).
+ * @param[out] aTtl          A pointer to an `uint32_t` to output TTL for the address. It can be NULL if caller does not
+ *                           want to get the TTL.
+ *
+ * @retval OT_ERROR_NONE       The address was read successfully.
+ * @retval OT_ERROR_NOT_FOUND  No address record in @p aResponse at @p aIndex.
+ * @retval OT_ERROR_PARSE      Could not parse the records in the @p aResponse.
+ *
+ */
+otError otDnsAddressResponseGetAddress(const otDnsAddressResponse *aResponse,
+                                       uint16_t                    aIndex,
+                                       otIp6Address *              aAddress,
+                                       uint32_t *                  aTtl);
+
+/**
+ * This type is an opaque representation of a response to a browse (service instance enumeration) DNS query.
+ *
+ * Pointers to instance of this type are provided from callback `otDnsBrowseCallback`.
+ *
+ */
+typedef struct otDnsBrowseResponse otDnsBrowseResponse;
+
+/**
+ * This function pointer is called when a DNS response is received for a browse (service instance enumeration) query
+ *
+ * Within this callback the user can use `otDnsBrowseResponseGet{Item}()` functions along with the @p aResponse
+ * pointer to get more info about the response.
+ *
+ * The @p aResponse pointer can only be used within this callback and after returning from this function it will not
+ * stay valid, so the user MUST NOT retain the @p aResponse pointer for later use.
+ *
+ * @param[in]  aError     The result of the DNS transaction.
+ * @param[in]  aResponse  A pointer to the response (it is always non-NULL).
+ * @param[in]  aContext   A pointer to application-specific context.
+ *
+ * For the full list of possible values for @p aError, please see `otDnsAddressCallback()`.
+ *
+ */
+typedef void (*otDnsBrowseCallback)(otError aError, const otDnsBrowseResponse *aResponse, void *aContext);
+
+/**
+ * This structure provides info for a DNS service instance.
+ *
+ */
+typedef struct otDnsServiceInfo
+{
+    uint32_t     mTtl;                ///< Service record TTL (in seconds).
+    uint16_t     mPort;               ///< Service port number.
+    uint16_t     mPriority;           ///< Service priority.
+    uint16_t     mWeight;             ///< Service weight.
+    char *       mHostNameBuffer;     ///< Buffer to output the service host name (can be NULL if not needed).
+    uint16_t     mHostNameBufferSize; ///< Size of `mHostNameBuffer`.
+    otIp6Address mHostAddress;        ///< The host IPv6 address. Set to all zero if not available.
```

Comment:
Added this to make it easier for user's of the API, so that they can get all service info in one call.

In particular, this allows the case where the user may not be interested in getting the host name and just want the IPv6 address of service instance. In this case, use can set `mHostNameBuffer` to `NULL` (no need for caller to allocate and a have a buffer to retrieve the host name).

`otDnsBrowseResponseGetHostAddress()` is there so we can handle the rare case where we may get multiple address (AAAA records) for the same host name.